src/jvm/main/org/jetbrains/kotlinx/lincheck/execution/ExecutionResult.kt

```diff
@@ -49,6 +49,53 @@ data class ExecutionResult(
     constructor(initResults: List<Result>, parallelResultsWithClock: List<List<ResultWithClock>>, postResults: List<Result>) :
         this(initResults, null, parallelResultsWithClock, null, postResults, null)
 
+    /**
+     * Number of threads with results.
+     */
+    val nThreads: Int = parallelResultsWithClock.size
+
+    private val initResultsWithClock: List<ResultWithClock> get() =
+        initResults.mapIndexed { i, result ->
+            val clock = emptyClock(nThreads).apply {
+                clock[INIT_THREAD_ID] = i
+            }
+            ResultWithClock(result, clock)
+        }
+
+    private val postResultsWithClock: List<ResultWithClock> get() =
+        postResults.mapIndexed { i, result ->
+            val clock = emptyClock(nThreads).apply {
+                for (iThread in 0 until nThreads) {
+                    clock[iThread] = when (iThread) {
+                        0 -> initResults.size + parallelResultsWithClock[0].size + i
+                        else -> parallelResultsWithClock[iThread].size
+                    }
+                }
+            }
+            ResultWithClock(result, clock)
+        }
+
+    /**
+     * List containing for each thread its list of results.
+     */
+    val threadsResultsWithClock: List<List<ResultWithClock>> = (0 until nThreads).map { i ->
+        val resultsWithUpdatedClock = parallelResultsWithClock[i].map { (result, clockOnStart) ->
+            val clock = emptyClock(nThreads).apply {
+                for (iThread in 0 until nThreads) {
+                    clock[iThread] = when (iThread) {
+                        0 -> initResults.size + clockOnStart.clock[0]
+                        else -> clockOnStart.clock[iThread]
+                    }
+                }
+            }
+            ResultWithClock(result, clock)
+        }
+        if (i == 0)
+            initResultsWithClock + resultsWithUpdatedClock + postResultsWithClock
```

Comment:
Here, the model is the init part is in the 0-th thread and the post part is in the (n+1)-th thread. Let's pack both init and post parts in the 1st thread, according to the updated mental model?

Reply:
Sorry, I didn't understand what do you mean here. 

> Here, the model is the init part is in the 0-th thread and the post part is in the (n+1)-th thread.

No, here I construct results representation with init/post parts executed in the 1st thread. 
I also update the happens-before clocks with respect to this representation.

Reply:
As I see, `initResultsWithClock` is indexed by `0` in the resulting list, and `postResultsWithClock` is indexed by `n+1` in the resulting list. Have I missed something?

Reply:
If we speak about `threadsResultsWithClock` list then by index `0` we have `initResultsWithClock + resultsWithUpdatedClock + postResultsWithClock`, that is init + 0th parallel thread + post.

Reply:
Note that `threadsResultsWithClock` starts with the following line:
```kotlin
val threadsResultsWithClock: List<List<ResultWithClock>> = (0 until nThreads).map { i -> ... 
```

Thus it is a list of `0 ... nThreads`. 

Then for each thread we rebuild its clocks:
```kotlin
val resultsWithUpdatedClock = parallelResultsWithClock[i].map { (result, clockOnStart) -> ... 
```

Finally, for `0th` thread we merge init + 0th thread in parallel part + post, for other threads we just take their results with updated clocks:

```kotlin
if (i == 0)
            initResultsWithClock + resultsWithUpdatedClock + postResultsWithClock
        else
            resultsWithUpdatedClock
```

Reply:
Would renaming `resultsWithUpdatedClock` to `threadResultsWithUpdatedClock` help here to understand the code?

Reply:
Putting `(0 until nThreads).map` on a new line and adding a comment would help. Also, please leave a TODO indicating that we need to refactor this part.

Reply:
Done